File: src/components/pipeline/DataValidationService.ts

import { ValidationRule } from './types';

/**
 * Generates PySpark code for data validation based on the provided rules
 * @param dataframeName The name of the dataframe to validate
 * @param rules Array of validation rules
 * @returns Generated PySpark code for validation
 */
export const generateValidationCode = (dataframeName: string, rules: ValidationRule[]): string => {
  let code = `# Data Validation for ${dataframeName}\n`;
  code += `from pyspark.sql import functions as F\n\n`;
  code += `# Original dataframe\n`;
  code += `original_df = ${dataframeName}\n\n`;
  code += `# Validation results\n`;
  code += `validation_results = []\n\n`;

  rules.forEach((rule, index) => {
    switch (rule.type) {
      case 'null':
        code += `# Check for nulls in ${rule.column}\n`;
        code += `null_count = ${dataframeName}.filter(F.col("${rule.column}").isNull()).count()\n`;
        code += `validation_results.append({"rule": "null_check", "column": "${rule.column}", "valid": null_count == 0, "failures": null_count})\n\n`;
        break;
      
      case 'dataType':
        code += `# Check data type for ${rule.column}\n`;
        code += `try:\n`;
        code += `    # Attempt to cast\n`;
        code += `    ${dataframeName}.select(F.col("${rule.column}").cast("${rule.value}")).filter(F.col("${rule.column}").isNotNull())\n`;
        code += `    validation_results.append({"rule": "data_type", "column": "${rule.column}", "valid": True, "failures": 0})\n`;
        code += `except Exception as e:\n`;
        code += `    validation_results.append({"rule": "data_type", "column": "${rule.column}", "valid": False, "failures": str(e)})\n\n`;
        break;
      
      case 'range':
        if (rule.condition && rule.value) {
          code += `# Check range for ${rule.column}\n`;
          code += `range_failures = ${dataframeName}.filter(~(F.col("${rule.column}") ${rule.condition} ${rule.value})).count()\n`;
          code += `validation_results.append({"rule": "range", "column": "${rule.column}", "valid": range_failures == 0, "failures": range_failures})\n\n`;
        }
        break;
      
      case 'regex':
        if (rule.value) {
          code += `# Check regex pattern for ${rule.column}\n`;
          code += `regex_failures = ${dataframeName}.filter(~F.regexp_extract(F.col("${rule.column}"), r'${rule.value}', 0).isNotNull()).count()\n`;
          code += `validation_results.append({"rule": "regex", "column": "${rule.column}", "valid": regex_failures == 0, "failures": regex_failures})\n\n`;
        }
        break;
      
      case 'schema':
        code += `# Schema validation\n`;
        code += `expected_schema_fields = ${JSON.stringify(rule.value)}\n`;
        code += `actual_fields = [field.name for field in ${dataframeName}.schema.fields]\n`;
        code += `missing_fields = [f for f in expected_schema_fields if f not in actual_fields]\n`;
        code += `validation_results.append({"rule": "schema", "valid": len(missing_fields) == 0, "failures": missing_fields})\n\n`;
        break;
      
      case 'custom':
        if (rule.condition) {
          code += `# Custom validation: ${rule.condition}\n`;
          code += `try:\n`;
          code += `    custom_result = ${dataframeName}.filter(${rule.condition}).count() > 0\n`;
          code += `    validation_results.append({"rule": "custom", "condition": "${rule.condition}", "valid": custom_result, "failures": 0 if custom_result else 1})\n`;
          code += `except Exception as e:\n`;
          code += `    validation_results.append({"rule": "custom", "condition": "${rule.condition}", "valid": False, "failures": str(e)})\n\n`;
        }
        break;
    }
  });

  code += `# Create a validation summary dataframe\n`;
  code += `validation_summary = spark.createDataFrame(validation_results)\n`;
  code += `display(validation_summary)\n\n`;
  code += `# Flag records that fail validation\n`;
  code += `# This creates a new column 'validation_status' with 'pass' or 'fail'\n`;
  code += `validated_df = ${dataframeName}\n\n`;
  
  // Add example of how to filter invalid records
  code += `# Example of how to flag or filter invalid records based on specific rules\n`;
  if (rules.length > 0 && rules[0].column) {
    const sampleRule = rules[0];
    code += `# Example: Flag records failing the ${sampleRule.type} check on ${sampleRule.column}\n`;
    if (sampleRule.type === 'null') {
      code += `validated_df = validated_df.withColumn("validation_status", F.when(F.col("${sampleRule.column}").isNull(), "fail").otherwise("pass"))\n\n`;
    } else if (sampleRule.type === 'range' && sampleRule.condition && sampleRule.value) {
      code += `validated_df = validated_df.withColumn("validation_status", F.when(F.col("${sampleRule.column}") ${sampleRule.condition} ${sampleRule.value}, "pass").otherwise("fail"))\n\n`;
    }
  } else {
    code += `# Add your validation flagging logic here based on your specific rules\n\n`;
  }

  code += `# Display validation results\n`;
  code += `display(validated_df)\n`;
  
  return code;
};

/**
 * Generates a simplified code summary for display in the node
 * @param rules Array of validation rules
 * @returns Simplified code summary
 */
export const generateValidationCodeSummary = (rules: ValidationRule[]): string => {
  if (rules.length === 0) {
    return "# No validation rules defined";
  }
  
  return `# Validating ${rules.length} rules:\n` + 
    rules.slice(0, 3).map(rule => 
      `# - ${rule.type} check${rule.column ? ` on "${rule.column}"` : ''}`
    ).join('\n') +
    (rules.length > 3 ? `\n# ... and ${rules.length - 3} more` : '');
};

File: python/services/data_validation_service.py
def generate_validation_code(dataframe_name, rules):
    """
    Generates PySpark code for data validation based on the provided rules
    
    Args:
        dataframe_name: The name of the dataframe to validate
        rules: Array of validation rules
        
    Returns:
        Generated PySpark code for validation
    """
    if not rules or not dataframe_name:
        return "# No validation rules provided"
        
    code = f"# Data Validation for {dataframe_name}\n"
    code += "from pyspark.sql import functions as F\n\n"
    code += f"# Original dataframe\n"
    code += f"original_df = {dataframe_name}\n\n"
    code += "# Validation results\n"
    code += "validation_results = []\n\n"
    
    for i, rule in enumerate(rules):
        rule_type = rule.get('type')
        column = rule.get('column')
        
        if not rule_type or not column:
            continue
            
        code += f"# Rule {i+1}: {rule_type} check on {column}\n"
        
        if rule_type == 'notNull':
            code += f"null_count = original_df.filter(F.col('{column}').isNull()).count()\n"
            code += f"validation_results.append({{\n"
            code += f"    'rule': 'Not Null Check',\n"
            code += f"    'column': '{column}',\n"
            code += f"    'passed': null_count == 0,\n"
            code += f"    'details': f'{{null_count}} null values found'\n"
            code += f"}})\n\n"
        elif rule_type == 'unique':
            code += f"total_count = original_df.count()\n"
            code += f"distinct_count = original_df.select('{column}').distinct().count()\n"
            code += f"validation_results.append({{\n"
            code += f"    'rule': 'Uniqueness Check',\n"
            code += f"    'column': '{column}',\n"
            code += f"    'passed': total_count == distinct_count,\n"
            code += f"    'details': f'{{total_count - distinct_count}} duplicate values found'\n"
            code += f"}})\n\n"
        elif rule_type == 'range':
            min_val = rule.get('min')
            max_val = rule.get('max')
            code += f"out_of_range = original_df.filter((F.col('{column}') < {min_val}) | (F.col('{column}') > {max_val})).count()\n"
            code += f"validation_results.append({{\n"
            code += f"    'rule': 'Range Check',\n"
            code += f"    'column': '{column}',\n"
            code += f"    'passed': out_of_range == 0,\n"
            code += f"    'details': f'{{out_of_range}} values outside range [{min_val}, {max_val}]'\n"
            code += f"}})\n\n"
    
    code += "# Create a validation summary dataframe\n"
    code += "validation_df = spark.createDataFrame(validation_results)\n"
    code += "display(validation_df)\n\n"
    code += "# Return validation status\n"
    code += "all_passed = all(result['passed'] for result in validation_results)\n"
    code += "print('Validation ' + ('passed' if all_passed else 'failed'))\n"
    
    return code

def generate_validation_code_summary(dataframe_name, rules):
    """Generate a summary of the validation code"""
    rule_count = len(rules) if rules else 0
    return f"Data validation for {dataframe_name} with {rule_count} rules"

def validate_data(dataframe, rules):
    """
    Validates a dataframe against a set of rules
    
    Args:
        dataframe: The dataframe to validate (or dataframe name if using in notebook context)
        rules: Array of validation rules
        
    Returns:
        Dictionary with validation results
    """
    # This is a placeholder function that would normally execute validation
    # In a real implementation, this would run the validation logic against a live dataframe
    # For now, we'll just return a mock result
    
    results = {
        'passed': True,
        'rules_checked': len(rules) if rules else 0,
        'rules_passed': len(rules) if rules else 0,
        'details': []
    }
    
    # In a real implementation, we would loop through the rules and check each one
    # against the dataframe
    
    return results